Put together a `net` module

joneshf opened the pull request:
This throws together a first pass at binding to the node.js `net` module. I tried to follow the style of the other packages we have, but let me know if I'm missing anything.

The commit has some comments, but I'm going to make a review so they're contextual.

Also, how do we enable travis for this repo? I don't see a way to turn on the app in the settings, and I don't have access through travis-ci.org.

joneshf committed:
Put together a `net` module

The changes here bind to the node.js `net` module.
The API tries to stay true to the node.js API while still being usable.

There are a couple of places where things are a bit inconsistent.
The naming for `onCloseServer` and `onCloseSocket`,
`onErrorServer` and `onErrorSocket`, and
`localAddressOption` and `localPortOption`.
This might be showing that there is a split waiting to happen.
Maybe we should have `Node.Net`, `Node.Net.Server` and `Node.Net.Socket`?
Maybe it doesn't matter that much?

In any case, this gives us bindings to the `net` module!

joneshf commented on src/Node/Net.purs:
```diff
@@ -0,0 +1,585 @@
+  , listenTCP
+  , listening
+  , localAddress
+  , localAddressOption
```
Most of the other options use the name of the field (like `fd` and `host` and whatever), since there's already a `localAddress` value, this had to be a different name. Is this an okay suffix? Should we suffix all options? Should we move the options elsewhere? Should we rename `localAddress` to something and make this option called `localAddress`? Should we do something else?

hdgarrood replied:
How about having a prefix on all option values which depends on the type of `Options` we are dealing with? So for example we'd have `connectLocalAddress` and `connectLocalPort`, but also `connectFamily`, `connectPort`, `connectHost`, and so on (for uniformity).

joneshf replied:
Works for me.

joneshf commented on src/Node/Net.purs:
```diff
@@ -0,0 +1,585 @@
+  , localAddress
+  , localAddressOption
+  , localPort
+  , localPortOption
```
Similar questions to `localAddress`.

joneshf commented on src/Node/Net.purs:
```diff
@@ -0,0 +1,585 @@
+  , localAddressOption
+  , localPort
+  , localPortOption
+  , onCloseServer
```
`onCloseServer` and `onCloseSocket` are suffixed because both the `Server` and the `Socket` have `'close'` events. Are these suffixes okay? Should we have separate modules: `Node.Net.Server`, `Node.Net.Socket` and each have an `onClose`? Should we do something else?

hdgarrood replied:
I think having a `Node.Net.Server` and a `Node.Net.Socket` module sounds good; this module is quite big as-is.

joneshf commented on src/Node/Net.purs:
```diff
@@ -0,0 +1,585 @@
+  , onData
+  , onDrain
+  , onEnd
+  , onErrorServer
```
Similar questions as `onCloseServer`/`onCloseSocket`.

hdgarrood commented on test/Test/Main.purs:
```diff
@@ -0,0 +1,73 @@
+import Effect.Console (errorShow, infoShow, logShow)
+import Node.Buffer (toString)
+import Node.Encoding (Encoding(..))
+import Node.Net (address, close, createConnectionTCP, createServer, endString, listenTCP, localAddress, localPort, onCloseServer, onCloseSocket, onConnect, onConnection, onData, onDrain, onEnd, onErrorServer, onErrorSocket, onListening, onLookup, onReady, onTimeout, writeString)
```
Let's use a qualified import here; otherwise we just impose ceremony and increase the likelihood of merge conflicts.

hdgarrood commented on src/Node/Net.purs:
```diff
@@ -0,0 +1,585 @@
+
+-- | Attempts to return the bound address of a `Server`.
+-- |
+-- | If the `Server` is not listening, no `Nothing` is returned.
```
typo here?

hdgarrood commented on src/Node/Net.js:
```diff
@@ -0,0 +1,403 @@
+exports.createConnectionImpl = function (x) {
+  return function (callback) {
+    return function () {
+      return net.createConnection(x, function () {
```
This could be written as `net.createConnection(x, callback)` right? That would avoid the extra function allocation each time too I think.

joneshf replied:
I originally had it that way, but it was hard to know which callbacks took arguments and which didn't. Some of them weren't being invoked (the `Effect _` wasn't being run); the callback was passed two or three arguments in uncurried form where it was expecting those arguments curried and then to invoke the `Effect _` afterward.

Should this FFI be written with `Effect.Uncurried` instead? That seems like it would solve that problem and also remove the allocation. I hadn't seen `Effect.Uncurried` used in other FFI modules in this organization, so I wasn't sure if I should do it here.

hdgarrood replied:
Yes, I think that would be better. The main reason we aren't using `Effect.Uncurried` elsewhere in node.js api bindings is that many of the bindings predate `Effect.Uncurried` I think.

hdgarrood commented on src/Node/Net.js:
```diff
@@ -0,0 +1,403 @@
+exports.onCloseServer = function (server) {
+  return function (callback) {
+    return function () {
+      server.on("close", function () {
```
Same here, I think this can be `server.on("close", callback)`.

hdgarrood commented on src/Node/Net.purs:
```diff
@@ -0,0 +1,585 @@
+
+-- | Attempts to return the port a client is connecting on.
+-- | E.g. if a client connects from `80`,
+-- | the result would be `Just "80"`.
```
Is this a typo? I guess we want `Just 80` if it's a `Maybe Int`.

joneshf committed:
Prefix all options with a namespace for consistency

We had a couple of outliers where we had to suffix them with `Option`.
For consistency sake,
we switch the scheme to prefixing them with the type of option they are.

joneshf committed:
Convert all FFI calls to uncurried versions

From feedback,
we can avoid some allocations if we do the uncurrying explicitly.

This cleans up the FFI implementation a bit at the expense of the PS.
That seems like a fair trade-off since PS is easier to maintain than JS.

There are two bugs that got fixed in this change as well.

The first is that we were typing `Lookup` incorrectly.
The `family` is a `Maybe Int`, not a `Maybe String`.
This was silently failing but not being caught for whatever reason.
There is some kind of concurrency issue with all the callbacks.

The second is that for `Effect _`,
`ado` generates different code from `do`.
The difference here means we get the inefficient version of the code.
Rather than use `ado` and suffer the penalty,
we move over to `do` and allow for more efficient code.

joneshf committed:
Write fewer event handlers in the FFI

All of the event handlers had the same format,
and really were doing the same thing.
There's a reason that `EventEmitter#on` is a thing.
We can use that to our advantage and write less JS!

joneshf committed:
Fix some typos

joneshf committed:
Use qualified imports

Per review, using unqualified imports tends to lead to merge conflicts.
We can mitigate that by qualifying imports.

joneshf committed:
Separate out `Socket` to its own module

Per review,
we can drop the namespace on some values if we separate modules.

Having done the move,
it's not clear that the extra level of nesting was worth it.
It was one or two values that needed a suffix.
But now we've got a whole new module and another file of JS to maintain.

Oh well, we're here now.

joneshf committed:
Separate out `Server` to its own module

This is the second part of movind to modules.
Again, not sure this is worth it as we end up with more code to maintain
and a less clear API overall.
We should have left well-enough alone and had six out of 78 concepts in
this module inconsistent.

Oh well, maybe someone will find a better set of values in the future.

joneshf commented:
I think I hit all the comments. Mind giving it another look over?

Also, any idea how to get CI working?

hdgarrood commented:
I think CI will "just work" as soon as there's a `.travis.yml` file in `master`.

hdgarrood commented:
Scratch that, I just found the switch I needed to toggle in the travis ci web interface. It should build the next time you push. I'll re-review in a little bit.

joneshf committed:
Fix bug with `onDrain`

We interpreted the node.js documentation for the `'drain'` event.
Attaching the listener doesn't return a `Boolean`;
it returns the `Socket` like all other `#on` calls.
Since we're not returning the `Socket` for any other listeners,
we don't want to do it here.

garyb commented:
I've added a `GITHUB_TOKEN` to the travis env so it should auto publish to pursuit when a GH release is made too (the `.travis.yml` was already set up for that).

joneshf commented:
@hdgarrood Bumping in case this dropped off your plate.

hdgarrood commented on bower.json:
```diff
@@ -0,0 +1,30 @@
+    "purescript-transformers": "1.0.0 - 4"
+  },
+  "devDependencies": {
+    "purescript-console": "4.0.0 - 4"
```
This seems a little bit strange to me. Does this mean the same as `>= 4.0.0 < 5`? If so, I think I'd prefer that; it seems clearer.

garyb replied:
Using just `4` is an option too, as that also means `>= 4.0.0 < 5.0.0` according to https://jubianchi.github.io/semver-check/#/constraint/4

hdgarrood replied:
That also seems less obvious, and less workable if you want to say something like `>= 4.1.1 < 5.0.0`. I think using the `>= x.y.z < a.b.c` scheme everywhere is ideal, because it can be used for any range we might want to express, and because you'll never need to look it up to find out what it means.

joneshf replied:
That would be clearer, but I can't seem to get that syntax to work for me.

This is the error I get:
```
yarn run bower install
yarn run v1.13.0
$ /home/joneshf/programming/purescript-node/purescript-node-net/node_modules/.bin/bower install
bower purescript-console#>= 4.0.0 && < 5       not-cached https://github.com/purescript/purescript-console.git#>= 4.0.0 && < 5
bower purescript-console#>= 4.0.0 && < 5          resolve https://github.com/purescript/purescript-console.git#>= 4.0.0 && < 5
bower purescript-console#>= 4.0.0 && < 5     ENORESTARGET Tag/branch >= 4.0.0 && < 5 does not exist

Additional error details:
Available tags: v0.1.0, v0.1.0-rc.1, v0.1.1, v1.0.0, v1.0.0-rc.1, v2.0.0, v3.0.0, v4.0.0, v4.1.0, v4.2.0
Available branches: master
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
Maybe I'm doing something wrong. If you can make it work, I'm more than happy to use it.

hdgarrood replied:
Try removing the `&&`?

joneshf replied:
Oh wow. That was it.

hdgarrood reviewed (approved):
Looks great, thanks! I'd say this is ready for an initial release.

joneshf committed:
Use a different syntax for constraints

Per review, this syntax should be a bit clearer.

joneshf commented:
Thanks for the review!
